community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/IndexCursorProgressor.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.newapi;
+
+import org.neo4j.internal.kernel.api.LabelSet;
+import org.neo4j.values.storable.Value;
+
+public interface IndexCursorProgressor
+{
+    boolean next();
+
+    void close();
+
+    interface NodeValueCursor
```

Comment:
The define the 3 types of counter parts of the `IndexCursorProgressor` implementations. Where should they live? I think it would be possible to embed the `IndexCursorProgressor` logic directly into the cursors, but that would some at a cost of making these classes bigger, and it would also introduce some code redundancy.

Reply:
if you need to implement an interface to make sense out of it it just better not to do that. Do not think that is any criteria in any way.

@fickludd it looks like those 3 Cursor classes have something in common - `initialization`and `done` parts that can be extracted into some sort of closeable cursor interface. 
For me, those 3 cursors look like some kind of Value cursors and IndexCursorProgressor would be the last place where I will gonna take a look into what kind of cursors do we have in case i want to implement new kind of cursor and make it comfortable in the existing  hierarchy. I would put them completely in a separate on in some sort of cursors helper class.

Reply:
Had a lengthy discussion with @burqen about this interface. The `Cursor` interfaces here really have nothing to do with cursors, so we renamed them. Added a lot of javadocs. This is really an internal index interface, so I also moved it next to `IndexReader` to signal that.